synapse/module_api/__init__.py

```diff
@@ -320,6 +322,59 @@ def get_state_events_in_room(
         state = yield defer.ensureDeferred(self._store.get_events(state_ids.values()))
         return state.values()
 
+    async def create_and_send_event_into_room(
+        self,
+        sender: str,
+        room_id: str,
+        event_type: str,
+        content: JsonDict,
+        state_key: Optional[str] = None,
+    ) -> EventBase:
+        """Create and send an event into a room.
+
+        Args:
+            sender: The user ID of the event sender.
+            room_id: The ID of the room to send the event into.
+            event_type: The type of the event.
+            content: The event content.
+            state_key: The event's state key. None if this is not a state event.
+
+        Returns:
+            The event that was sent.
+
+        Raises:
+            SynapseError if the event was not allowed.
+        """
+        if event_type == EventTypes.Member:
+            raise Exception(
+                "Synapse modules are not currently able to send m.room.member events"
+            )
+
+        # Build event dictionary
+        event_dict = {
+            "type": event_type,
+            "content": content,
+            "room_id": room_id,
+            "sender": sender,
+        }
+
+        # Check whether this is intended to be a state event
+        if state_key is not None:
+            event_dict["state_key"] = state_key
```

Comment:
any reason not to put responsibility for this on the caller, and just have it pass in an event dict?

Reply:
Not really I suppose, other than a slightly simpler API surface. I suppose doing so might give the caller even greater flexibility in defining events though?

Reply:
imho a single `dict` is a simpler api than a dict and a type and a room id and a sender.

flexibility ftw

keeping ModuleApi as a very thin layer also ftw

Reply:
Heh, fair points all around. I've simplified it quite a bit in https://github.com/matrix-org/synapse/pull/8479/commits/16f979c1b58e7505f7e1a5312aa7510eb5d64117